pkg/hostpath/controllerserver.go

```diff
@@ -139,3 +142,151 @@ func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req
+	hostPathVolume, ok := hostPathVolumes[volumeID]
+	if !ok {
+		return nil, status.Error(codes.Internal, "volumeID is not exist")
+	}
+
+	snapshotID := uuid.NewUUID().String()
+	createAt := time.Now().UnixNano()
+	volPath := hostPathVolume.VolPath
+	file := snapshotRoot + snapshotID + ".tgz"
+	args := []string{"czf", file, "-C", volPath, "."}
+	executor := utilexec.New()
+	out, err := executor.Command("tar", args...).CombinedOutput()
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v: %s", err, out))
+	}
+
+	glog.V(4).Infof("create volume snapshot %s", file)
+	snapshot := hostPathSnapshot{}
+	snapshot.Name = req.GetName()
+	snapshot.Id = snapshotID
+	snapshot.VolID = volumeID
+	snapshot.Path = file
+	snapshot.CreateAt = createAt
+	snapshot.Status = &csi.SnapshotStatus{
+		Type:    csi.SnapshotStatus_READY,
+		Details: fmt.Sprint("Successfully create snapshot"),
+	}
+
+	hostPathVolumeSnapshots[snapshotID] = snapshot
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			Id:             snapshot.Id,
+			SourceVolumeId: snapshot.VolID,
+			CreatedAt:      snapshot.CreateAt,
+			Status:         snapshot.Status,
+		},
+	}, nil
+}
+
+func (cs *controllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	// Check arguments
+	if len(req.GetSnapshotId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot ID missing in request")
+	}
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid delete snapshot req: %v", req)
+		return nil, err
+	}
+	snapshotID := req.SnapshotId
+	glog.V(4).Infof("deleting volume %s", snapshotID)
+	path := snapshotRoot + snapshotID + ".tgz"
+	os.RemoveAll(path)
+	delete(hostPathVolumeSnapshots, snapshotID)
+	return &csi.DeleteSnapshotResponse{}, nil
+}
+
+func (cs *controllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_LIST_SNAPSHOTS); err != nil {
+		glog.V(3).Infof("invalid list snapshot req: %v", req)
+		return nil, err
+	}
+
+	var snapshots []hostPathSnapshot
+
+	// case 1: SnapshotId is not empty, return snapshots that match the snapshot id.
+	if len(req.GetSnapshotId()) != 0 {
+		snapshotID := req.SnapshotId
+		if snapshot, ok := hostPathVolumeSnapshots[snapshotID]; ok {
+			snapshots = append(snapshots, snapshot)
+			// jump to the result processing.
+			goto result
+		}
+	}
+
+	// case 2: SourceVolumeId is not empty, return snapshots that match the source volume id.
+	if len(req.SourceVolumeId) != 0 {
+		for _, snapshot := range hostPathVolumeSnapshots {
+			if snapshot.VolID == req.SourceVolumeId {
+				snapshots = append(snapshots, snapshot)
+				// jump to the result processing.
+				goto result
+			}
+		}
+	}
```

Comment:
Shouldn't there be duplicate snapshot entries where both, snapshotId and sourceVolumeId are given?

Reply:
Actually it won't as it'll goto result once process this block. *You can skip this comment*